apps/ejabberd/src/ejabberd_c2s.erl

```diff
@@ -939,7 +951,20 @@ process_outgoing_stanza(El, StateData) ->
                 _ ->
                     NewEl1
             end,
-    NState = process_outgoing_stanza(ToJID, Name, {Attrs, NewEl, FromJID, StateData, Server, User}),
+    Acc2 = mongoose_acc:put(element, NewEl, Acc1),
+    Name = mongoose_acc:get(name, Acc2),
+    NState = case Name of
+        <<"presence">> ->
+            % new-style
+            % I think it still makes sense to pass all those arguments, for performance reasons
```

Comment:
ad.1: Frankly speaking, I have no idea how much it costs to read a value from a map, compared to passing value through a function call. Since those values are already there, I though it'd be a bit cheaper to pass them along. If you think the cost is meaningless then yes, we can stuff everything into acc and pass only values which are required for pattern matching. Though, if we begin to rely on state data in accumulator, then we'll probably need to remember to update it when changed, and remove it before sending the acc to sm.